### PR TITLE
[#113]: create chain segment factory

### LINKED ISSUE #113: [Toggle Core] Create ChainSegmentFactory
**Epic:** https://github.com/pheature-flags/pheature-flags/issues/1 Toggle Core
**Package:** https://github.com/pheature-flags/toggle-core 

**Description:**

Create ChainSegmentFactory class implementing the interface defined in issue https://github.com/pheature-flags/pheature-flags/issues/111

The chain segment factory should receive a list of segment factory interfaces. Then it should be able to use a correct factory depending on the given input value. Also, it should return a list of available Segment types.

### PR DESCRIPTION
**Pull request by kpicaza**
Creates ChainSegmentFactory as described in issue #113

### CODE REVIEW

File: packages/toggle-core/test/Read/ChainSegmentFactoryTest.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        $segmentFactory->expects(static::once())
+            ->method('types')
+            ->willReturn([self::SEGMENT_TYPE]);
+        $segmentFactory->expects(static::once())
+            ->method('create')
+            ->with(self::SEGMENT_ID, self::SEGMENT_TYPE, self::PAYLOAD)
+            ->willReturn($this->createMock(Segment::class));
+        $chainSegmentFactory = new ChainSegmentFactory($segmentFactory);
+
+        $chainSegmentFactory->create(self::SEGMENT_ID, self::SEGMENT_TYPE, self::PAYLOAD);
```
Comment (xserrat) [inline packages/toggle-core/test/Read/ChainSegmentFactoryTest.php]: In order to add an extra assert, we can declare a variable assigning the mocked segment that should be the same as the returned by the `ChainSegmentFactory`:

```suggestion
        $expectedSegment = $this->createMock(Segment::class);
        $segmentFactory->expects(static::once())
            ->method('create')
            ->with(self::SEGMENT_ID, self::SEGMENT_TYPE, self::PAYLOAD)
            ->willReturn($expectedSegment);
        $chainSegmentFactory = new ChainSegmentFactory($segmentFactory);

        $actual = $chainSegmentFactory->create(self::SEGMENT_ID, self::SEGMENT_TYPE, self::PAYLOAD);

        self::assertSame($expectedSegment, $actual);
```

What do you think?

**Review by xserrat — APPROVED** [review]
🚀🚀 Only a suggestion to add an extra assertion. Good work!!

File: packages/toggle-core/test/Read/ChainSegmentFactoryTest.php (hunk 1 above)
Comment (xserrat) [reply packages/toggle-core/test/Read/ChainSegmentFactoryTest.php]: Regarding this change we could add the same logic for the changes of this PR: https://github.com/pheature-flags/pheature-flags/pull/158

Comment (kpicaza) [reply packages/toggle-core/test/Read/ChainSegmentFactoryTest.php]: Thanks @xserrat ;-D I will make both changes :rocket::rocket::rocket:

Comment (kpicaza) [reply packages/toggle-core/test/Read/ChainSegmentFactoryTest.php]: done;-D

**Pull request merged**